### DISCUSSION TITLE
Allow automatic inclusion of `COMMENT ON` statements when calling `MetaData.create_all()`

### DISCUSSION THREAD

**Discussion by zaneselvans** [sqlalchemy/sqlalchemy#11656] [Ideas]
To ensure that descriptive metadata describing tables and columns can easily be included in the database if they're present, it would be helpful if `MetaData.create_all()` had the ability to also generate appropriate `COMMENT ON` statements if the tables and columns described have associated comments.

This would make it easier to generate a database schema / SQL file that can be used to programmatically build documentation for the database.

My current workaround looks like:

```python
import sqlalchemy as sa
import duckdb

duckdb_path = "/path/to/database.duckdb"
metadata = sa.MetaData()
engine = sa.create_engine(f"duckdb:///{duckdb_path}")
metadata.reflect(bind=engine)
metadata.create_all(engine)

with duckdb.connect(database=str(duckdb_path)) as duckdb_conn:
    duckdb_cursor = duckdb_conn.cursor()
    # Iterate through the tables in order of foreign key dependency
    # Explicitly add comments to all tables and columns, since this is not done
    # automatically by create_all():
    for table in metadata.sorted_tables:
        duckdb_cursor.execute(str(sa.schema.SetTableComment(table)))
        for column in table.columns:
            duckdb_cursor.execute(str(sa.schema.SetColumnComment(column)))
```

But I could imagine it just being something like:

```python
import sqlalchemy as sa

duckdb_path = "/path/to/database.duckdb"
metadata = sa.MetaData()
engine = sa.create_engine(f"duckdb:///{duckdb_path}")
metadata.reflect(bind=engine)
metadata.create_all(engine, comments=True)
```

(Assuming the database supports the inclusion of comments)

See also https://github.com/duckdb/duckdb/discussions/13178

**Comment by zzzeek** [comment]
metadata.create_all() already has this capability if I'm understanding everything here correctly, using the [comment](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Table.params.comment) parameter on Table and Column:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import Table

m = MetaData()

t = Table("t", m, Column("x", Integer, comment="column x"), comment="table t")

e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)

m.create_all(e)

```

output:

```
CREATE TABLE t (
	x INTEGER
)


2024-07-28 22:54:24,350 INFO sqlalchemy.engine.Engine [no key 0.00005s] {}
2024-07-28 22:54:24,351 INFO sqlalchemy.engine.Engine COMMENT ON TABLE t IS 'table t'
2024-07-28 22:54:24,351 INFO sqlalchemy.engine.Engine [no key 0.00005s] {}
2024-07-28 22:54:24,353 INFO sqlalchemy.engine.Engine COMMENT ON COLUMN t.x IS 'column x'
2024-07-28 22:54:24,353 INFO sqlalchemy.engine.Engine [no key 0.00007s] {}
2024-07-28 22:54:24,354 INFO sqlalchemy.engine.Engine COMMIT

```

seems to be what you're looking for, this feature has been present since version 1.2 many years ago

**Reply by zaneselvans** [reply]
Hmm. I was surprised that the comments didn't come through. Maybe support for passing the comments through just hasn't been implemented in the DuckDB driver?

```python
m = MetaData()
t = Table("t", m, Column("x", Integer, comment="column x"), comment="table t")
e = create_engine("duckdb:///:memory:", echo=True)
m.create_all(e)
```

No, that's not it:
```
CREATE TABLE t (
	x INTEGER
)

2024-07-28 22:04:57,585 INFO sqlalchemy.engine.Engine [no key 0.00052s] ()
2024-07-28 22:04:57,586 INFO sqlalchemy.engine.Engine COMMENT ON TABLE t IS 'table t'
2024-07-28 22:04:57,587 INFO sqlalchemy.engine.Engine [no key 0.00026s] ()
2024-07-28 22:04:57,587 INFO sqlalchemy.engine.Engine COMMENT ON COLUMN t.x IS 'column x'
2024-07-28 22:04:57,587 INFO sqlalchemy.engine.Engine [no key 0.00020s] ()
2024-07-28 22:04:57,588 INFO sqlalchemy.engine.Engine COMMIT
```

Oooh. It looks like there are a bunch of columns without comments that aren't mine (they're just part of DuckDB) that happened to be at the beginning and the end of the list of columns, so they all looked like they were NULL when I selected all with

```sql
SELECT comment FROM duckdb_columns();
```

But this shows my columns and the expected comments.
```sql
SELECT table_name, column_name, data_type, is_nullable, comment
FROM duckdb_columns()
WHERE comment IS NOT NULL;
```

And I coincidentally had a bug elsewhere that was preventing the comment from being added to the tables. And together these made it seem like the comments were just not coming through at all. So it all totally works.

Sorry for the distraction and thanks for responding so quickly!
